File: java/femtozip/src/org/toubassi/femtozip/models/OptimizingCompressionModel.java
/**
 *   Copyright 2011 Garrick Toubassi
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
package org.toubassi.femtozip.models;

import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

import org.toubassi.femtozip.CompressionModel;
import org.toubassi.femtozip.DocumentList;

public class OptimizingCompressionModel extends CompressionModel {

    public static class CompressionResult implements Comparable<CompressionResult>{
        public CompressionModel model;
        public int totalCompressedSize;
        public int totalDataSize;

        public CompressionResult(CompressionModel model) {
            this.model = model;
        }

        public int compareTo(CompressionResult other) {
            return totalCompressedSize - other.totalCompressedSize;
        }
        
        public void accumulate(CompressionResult result) {
            totalCompressedSize += result.totalCompressedSize < result.totalDataSize ? result.totalCompressedSize:  result.totalDataSize;
            totalDataSize += result.totalDataSize;
        }
        
        public String toString() {
            DecimalFormat format = new DecimalFormat("#.##");
            return model.getClass().getSimpleName() + " " + format.format((100f * totalCompressedSize) / totalDataSize) + "% (" + totalCompressedSize + " from " + totalDataSize + " bytes)";
        }
    }

    private ArrayList<CompressionResult> results = new ArrayList<CompressionResult>();
    private ArrayList<CompressionResult> sortedResults;
    private int totalDataSize;

    public OptimizingCompressionModel() {
        this("DeflateFrequency,GZip,GZipDictionary,NibbleFrequency,OffsetNibbleFrequency,OffsetNibbleHuffman,PureArithCoding,PureHuffman,SplitFrequency,TripleNibbleFrequency,UnifiedFrequency,Noop,VariableInt");
    }
    
    public OptimizingCompressionModel(String modelNames) {
        this(modelNames.split(","));
    }
    
    public OptimizingCompressionModel(String[] modelNames) {
        for (String modelName : modelNames) {
            CompressionModel model = CompressionModel.instantiateCompressionModel(modelName);
            results.add(new CompressionResult(model));
        }
    }
    
    public void load(DataInputStream in) throws IOException {
        throw new UnsupportedOperationException();
    }
    
    public void save(DataOutputStream out) throws IOException {
        throw new UnsupportedOperationException("You should save the best performing model, not the OptimizingCompressionModel");
    }
    
    @Override
    public void build(DocumentList documents) throws IOException {
        buildDictionaryIfUnspecified(documents);
        
        for (CompressionResult result : results) {
            
            // No need to recompute this over and over.  This assumes all types of compression model
            // compute the dictionary the same way.
            result.model.setDictionary(getDictionary());
            
            result.model.build(documents);
        }
    }
    
    /**
     * Should be a different set of documents from those used to build
     */
    public void optimize(DocumentList documents) throws IOException {
        
        for (int i = 0, count = documents.size(); i < count; i++) {
            byte[] data = documents.get(i);
            
            totalDataSize += data.length;
            
            for (CompressionResult result : results) {
                ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
                result.model.compress(data, bytesOut);
                
                if (true) {
                    byte[] decompressed = result.model.decompress(bytesOut.toByteArray());
                    if (!Arrays.equals(data, decompressed)) {
                        throw new RuntimeException("Compress/Decompress round trip failed for " + result.model.getClass().getSimpleName());
                    }
                }
                
                result.totalCompressedSize += bytesOut.size();
                result.totalDataSize += data.length;
            }
        }
        
        sortedResults = new ArrayList<CompressionResult>(results);
        Collections.sort(sortedResults);
    }

    
    public void compress(byte[] data, OutputStream out) throws IOException {
        getBestPerformingModel().compress(data, out);
    }
    

    public byte[] decompress(byte[] compressedData) {
        return getBestPerformingModel().decompress(compressedData);
    }
    
    public CompressionModel getBestPerformingModel() {
        int best = sortedResults.get(0).totalCompressedSize;
        return best > totalDataSize ? null : sortedResults.get(0).model;
    }
    
    public CompressionResult getBestPerformingResult() {
        return sortedResults.get(0);
    }
    
    public void encodeLiteral(int aByte) {
        throw new UnsupportedOperationException();
    }

    public void encodeSubstring(int offset, int length) {
        throw new UnsupportedOperationException();
    }
    
    public void endEncoding() {
        throw new UnsupportedOperationException();
    }
   
    public void aggregateResults(ArrayList<CompressionResult> aggregateResults) {
        if (aggregateResults.size() == 0) {
            for (CompressionResult result : results) {
                aggregateResults.add(new CompressionResult(result.model));
            }
        }
        for (int i = 0, count = results.size(); i < count; i++) {
            CompressionResult result = results.get(i);
            CompressionResult aggregate = aggregateResults.get(i);
            
            if (result.model.getClass() != aggregate.model.getClass()){
                throw new RuntimeException("Can't aggregate across different sets of models");
            }
            aggregate.accumulate(result);
        }
    }
    
    public void dump() {
        if (totalDataSize <= 0) {
            return;
        }
        
        
        for (CompressionResult result : sortedResults) {
            System.out.println(result);
        }
    }
}

File: cpp/cpp-femtozip/src/OptimizingCompressionModel.cpp
/**
 *   Copyright 2011 Garrick Toubassi
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
/*
 * OptimizingCompressionModel.cpp
 *
 *  Created on: Mar 8, 2011
 *      Author: gtoubassi
 */

#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include "OptimizingCompressionModel.h"
#include "OffsetNibbleHuffmanCompressionModel.h"
#include "PureHuffmanCompressionModel.h"

using namespace std;

namespace femtozip {

OptimizingCompressionModel::OptimizingCompressionModel() {
    results.push_back(CompressionResult(new OffsetNibbleHuffmanCompressionModel()));
    results.push_back(CompressionResult(new PureHuffmanCompressionModel()));
}

OptimizingCompressionModel::OptimizingCompressionModel(vector<string>& models) {
    for (vector<string>::iterator i = models.begin(); i != models.end(); i++) {
        CompressionModel *model;

        if (*i == "PureHuffman") {
            model = new PureHuffmanCompressionModel();
        }
        else if (*i == "OFfsetNibbleHuffman") {
            model = new OffsetNibbleHuffmanCompressionModel();
        }
        else {
            throw "Unknown model";
        }
        CompressionResult result(model);
        results.push_back(result);
    }
}

OptimizingCompressionModel::~OptimizingCompressionModel() {
    for (vector<CompressionResult>::iterator i = results.begin(); i != results.end(); i++) {
        delete i->model;
    }
}

void OptimizingCompressionModel::build(DocumentList& documents) {
    buildDictionaryIfUnspecified(documents);

    for (vector<CompressionResult>::iterator result = results.begin(); result != results.end(); result++) {

        // No need to recompute this over and over.  This assumes all types of compression model
        // compute the dictionary the same way.
        int dictLength;
        const char *dict = getDictionary(dictLength);
        result->model->setDictionary(dict, dictLength);
        result->model->build(documents);
    }
}

void OptimizingCompressionModel::optimize(DocumentList& documents) {
    for (int i = 0, count = documents.size(); i < count; i++) {
        int length;
        const char *buf = documents.get(i, length);

        totalDataSize += length;

        for (vector<CompressionResult>::iterator result = results.begin(); result != results.end(); result++) {
            ostringstream compressOut;
            result->model->compress(buf, length, compressOut);
            string compressed = compressOut.str();

            if (true) {
                ostringstream decompressOut;
                result->model->decompress(compressed.c_str(), compressed.length(), decompressOut);
                string decompressed = decompressOut.str();
                if (decompressed.length() != static_cast<unsigned int>(length) || memcmp(decompressed.c_str(), buf, length) != 0) {
                    throw "Compress/Decompress roundtrip failed";
                }
            }

            result->totalCompressedSize += compressed.size();
            result->totalDataSize += length;
        }
    }

    sortedResults = results;
    sort(sortedResults.begin(), sortedResults.end());
}


void OptimizingCompressionModel::compress(const char *buf, int length, ostream& out) {
    getBestPerformingModel()->compress(buf, length, out);
}

void OptimizingCompressionModel::decompress(const char *buf, int length, ostream& out) {
    getBestPerformingModel()->decompress(buf, length, out);
}

CompressionModel *OptimizingCompressionModel::getBestPerformingModel() {
    int best = sortedResults[0].totalCompressedSize;
    return best > totalDataSize ? 0 : sortedResults[0].model;
}

OptimizingCompressionModel::CompressionResult &OptimizingCompressionModel::getBestPerformingResult() {
    return sortedResults[0];
}

void OptimizingCompressionModel::encodeLiteral(int aByte) {
    throw "OptimizingCompressionModel::encodeLiteral unsupported";
}

void OptimizingCompressionModel::encodeSubstring(int offset, int length) {
    throw "OptimizingCompressionModel::encodeSubstring unsupported";
}

void OptimizingCompressionModel::endEncoding() {
    throw "OptimizingCompressionModel::endEncoding unsupported";
}

void OptimizingCompressionModel::aggregateResults(vector<CompressionResult> &aggregateResults) {
    if (aggregateResults.size() == 0) {
        for (vector<CompressionResult>::iterator result = results.begin(); result != results.end(); result++) {
            CompressionResult r(result->model);
            aggregateResults.push_back(r);
        }
    }
    for (int i = 0, count = results.size(); i < count; i++) {
        CompressionResult& result = results[i];
        CompressionResult& aggregate = aggregateResults[i];

        aggregate.accumulate(result);
    }
}


}
